File: 2023/day03-1.hs
type Coord = (Int, Int)

type Component = (Int, [Coord])

solution :: String -> Int
solution content =
  sum
    . map fst
    . filter (\(_, coords) -> (any (`elem` symbols) . surroundings) coords)
    $ components
  where
    components = parseComponents content
    symbols = parseSymbols content

surroundings :: [Coord] -> [Coord]
surroundings [] = []
surroundings ((x, y) : xs) =
  [ (x - 1, y - 1),
    (x, y - 1),
    (x + 1, y - 1),
    (x - 1, y),
    (x + 1, y),
    (x - 1, y + 1),
    (x, y + 1),
    (x + 1, y + 1)
  ]
    ++ surroundings xs

parseSymbols :: String -> [Coord]
parseSymbols = helper 0 0
  where
    helper _ _ "" = []
    helper _ y ('\n' : cs) = helper 0 (y + 1) cs
    helper x y (c : cs)
      | c `elem` ".1234567890" = helper (x + 1) y cs
      | otherwise = (x, y) : helper (x + 1) y cs

parseComponents :: String -> [Component]
parseComponents = helper 0 0 (0, [])
  where
    helper _ _ _ "" = []
    helper _ y (_, []) ('\n' : cs) = helper 0 (y + 1) (0, []) cs
    helper _ y cmp ('\n' : cs) = cmp : helper 0 (y + 1) (0, []) cs
    helper x y (_, []) (c : cs)
      | c `elem` "1234567890" = helper (x + 1) y (read (c : ""), [(x, y)]) cs
      | otherwise = helper (x + 1) y (0, []) cs
    helper x y (v, w) (c : cs)
      | c `elem` "1234567890" = helper (x + 1) y (v * 10 + read (c : ""), (x, y) : w) cs
      | otherwise = (v, w) : helper (x + 1) y (0, []) cs

main :: IO ()
main = do readFile "day03.dat" >>= print . solution

File: 2023/day03_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring

from typing import Any

type Coord = tuple[int, int]
type Component = tuple[int, list[Coord]]


def solution(content: str) -> Any:
    components: list[Component] = parse_components(content)
    symbols: list[Coord] = parse_symbols(content)
    return sum(
        val
        for val, coords in components
        if [c for c in surroundings(coords) if c in symbols]
    )


def surroundings(coords: list[Coord]):
    return {
        c
        for x, y in coords
        for c in [
            (x - 1, y - 1),
            (x, y - 1),
            (x + 1, y - 1),
            (x - 1, y),
            (x + 1, y),
            (x - 1, y + 1),
            (x, y + 1),
            (x + 1, y + 1),
        ]
    }


def parse_symbols(content: str):
    return [
        (row_index, col_index)
        for row_index, row in enumerate(content.split("\n"))
        for col_index, char in enumerate(row)
        if char not in ".1234567890"
    ]


def parse_components(content: str):
    component: None | Component = None
    output = []
    for row_index, row in enumerate(content.split("\n")):
        for col_index, char in enumerate(row):
            if char in "1234567890":
                if component is None:
                    component = (int(char), [(row_index, col_index)])
                else:
                    value, coords = component
                    component = (
                        value * 10 + int(char),
                        [*coords, (row_index, col_index)],
                    )
            elif component:
                output.append(component)
                component = None
        if component:
            output.append(component)
            component = None
    return output


def main() -> None:
    with open("day03.dat", "r", encoding="utf8") as file:
        print(solution(file.read().strip()))


if __name__ == "__main__":
    main()
